applications/DEMApplication/python_scripts/DEM_analysis_stage.py

```diff
@@ -0,0 +1,647 @@
+                self.FillAnalyticSubModelPartsWithNewParticles()
+
+    def BeforePrintingOperations(self, time):
+        pass
+
+    def FinalizeSolutionStep(self):
+        super(DEMAnalysisStage, self).FinalizeSolutionStep()
+        self.AfterSolveOperations()
+
+        self.DEMFEMProcedures.MoveAllMeshes(self.all_model_parts, self.time, self.solver.dt)
+
+        ##### adding DEM elements by the inlet ######
+        if self.DEM_parameters["dem_inlet_option"].GetBool():
+            self.DEM_inlet.CreateElementsFromInletMesh(self.spheres_model_part, self.cluster_model_part, self.creator_destructor)  # After solving, to make sure that neighbours are already set.
+
+        stepinfo = self.report.StepiReport(timer, self.time, self.step)
+        if stepinfo:
+            self.KRATOSprint(stepinfo)
+
+    def OutputSolutionStep(self):
+        #### PRINTING GRAPHS ####
+        #-------os.chdir(self.graphs_path)
+        self.post_utils.ComputeMeanVelocitiesInTrap("Average_Velocity.txt", self.time, self.graphs_path)
+        self.materialTest.MeasureForcesAndPressure()
+        self.materialTest.PrintGraph(self.time)
+        self.DEMFEMProcedures.PrintGraph(self.time)
+        self.DEMFEMProcedures.PrintBallsGraph(self.time)
+        self.DEMEnergyCalculator.CalculateEnergyAndPlot(self.time)
+        self.BeforePrintingOperations(self.time)
+        self.PrintResults()
+        self.FinalizeTimeStep(self.time)
+
+    def AfterSolveOperations(self):
+        if self.post_normal_impact_velocity_option:
+            self.particle_watcher.MakeMeasurements(self.analytic_model_part)
+            if self.IsTimeToPrintPostProcess(self.time):
+                self.particle_watcher.SetNodalMaxImpactVelocities(self.analytic_model_part)
+                self.particle_watcher.SetNodalMaxFaceImpactVelocities(self.analytic_model_part)
+
+        #Phantom Walls
+        self.RunAnalytics(self.time, self.IsTimeToPrintPostProcess(self.time))
+
+    def FinalizeTimeStep(self, time):
+        pass
+
+    def BreakSolutionStepsLoop(self):
+        return False
+
+    def TheSimulationMustGoOn(self):
+        it_must_or_not = self.time < self.end_time
+        it_must_or_not = it_must_or_not and not self.BreakSolutionStepsLoop()
+        return it_must_or_not
+
+    def Finalize(self):
+
+        self.KRATOSprint("Finalizing execution...")
+        self.GraphicalOutputFinalize()
+        self.materialTest.FinalizeGraphs()
+        self.DEMFEMProcedures.FinalizeGraphs(self.rigid_face_model_part)
+        self.DEMFEMProcedures.FinalizeBallsGraphs(self.spheres_model_part)
+        self.DEMEnergyCalculator.FinalizeEnergyPlot()
+        self.CleanUpOperations()
+
+        #------os.chdir(self.main_path)
+
+    def CleanUpOperations(self):
+
+        self.procedures.DeleteFiles()
+
+        self.KRATOSprint(self.report.FinalReport(timer))
+
+        if self.post_normal_impact_velocity_option:
+            del self.analytic_model_part
+
+        del self.KRATOSprint
```

Comment:
is this needed?
This does not guarantee that the garbace-collector is being called (it's python after all ...)

Reply:
At one stage we had problems with memory leaks from python and this is the result of a trial-and-error process to solve them. The origin of the memory leaks was the existence of circular dependencies with member variables such as the ones being deleted here. We can study the matter to try to eliminate these deletions but in a future PR